lib/useOnyx.ts

```diff
@@ -0,0 +1,189 @@
+    return (Onyx.tryGetCachedValue(key, {selector}) ?? null) as CachedValue<TKey, TValue>;
+}
+
+function useOnyx<TKey extends OnyxKey, TReturnData = OnyxValue<TKey>>(key: TKey, options?: UseOnyxOptions<TKey, TReturnData>): UseOnyxData<TKey, TReturnData> {
+    const connectionIDRef = useRef<number | null>(null);
+    const previousKey = usePrevious(key);
+
+    /**
+     * Used to store collection objects or selected data since they aren't stored in the cache.
+     */
+    const currentDataRef = useRef<CachedValue<TKey, TReturnData> | null>(null);
+
+    const isFirstRenderRef = useRef(true);
+    const fetchStatusRef = useRef<FetchStatus>('loading');
+
+    useEffect(() => {
+        /**
+         * This condition will ensure we can only handle dynamic collection member keys.
+         */
+        if (previousKey === key || (isCollectionMemberKey(previousKey) && isCollectionMemberKey(key))) {
+            return;
+        }
+
+        throw new Error(
+            `'${previousKey}' key can't be changed to '${key}'. useOnyx() only supports dynamic keys if they are both collection member keys e.g. from 'collection_id1' to 'collection_id2'.`,
+        );
+    }, [previousKey, key]);
+
+    /**
+     * According to React docs, `getSnapshot` is a function that returns a snapshot of the data in the store that’s needed by the component.
+     * **While the store has not changed, repeated calls to getSnapshot must return the same value.**
+     * If the store changes and the returned value is different (as compared by Object.is), React re-renders the component.
+     */
+    const getSnapshot = useCallback(() => {
+        /**
+         * Case 1 - We have a non-collection key without selector
+         *
+         * We just return the data from the Onyx cache.
+         */
+        if (!Onyx.isCollectionKey(key) && !options?.selector) {
+            return getCachedValue(key);
+        }
+
+        /**
+         * Case 2 - We have a non-collection/collection key with/without selector
+         *
+         * Since both collection objects and selected data are not directly stored in the cache, we need to generate them with `getCachedValue`
+         * and deep compare with our previous internal data.
+         *
+         * If they are not equal, we update the internal data and return it.
+         *
+         * If they are equal, we just return the previous internal data.
+         */
+        const newData = getCachedValue(key, options?.selector);
+        if (!deepEqual(currentDataRef.current, newData)) {
+            currentDataRef.current = newData as CachedValue<TKey, TReturnData>;
+        }
+
+        return currentDataRef.current as CachedValue<TKey, TReturnData>;
+    }, [key, options?.selector]);
+
+    const subscribe = useCallback(
+        (onStoreChange: () => void) => {
+            connectionIDRef.current = Onyx.connect({
+                key: key as CollectionKeyBase,
```

Comment:
Why would this be a `CollectionKeyBase`? Couldn't it be any `OnyxKey` _or_ `CollectionKeyBase`?

Reply:
It's because I'm supplying a boolean to `waitForCollectionCallback` and that is making TS require a `CollectionKeyBase` here, but as this connect usage is generic it's fine to assert to `CollectionKeyBase` imo.